### PR TITLE
OCaml Bidirectional Records | Ksenia Melashenko 

### PR DESCRIPTION
**Pull request by ksenmel**
- Ast 
- Parser for primitive types, expressions etc.
- Trivial parser for type ascription and records 
- Factorial inline test

### CODE REVIEW

**Comment by Kakadu** [comment]
Документация и тестовое покрытие  должны скоро появиться.

https://kakadu.github.io/fp2023/docs/OCaml_BidirectionalRecords

https://kakadu.github.io/fp2023/cov/OCaml_BidirectionalRecords

2023-11-06 17:55

**Comment by Kakadu** [comment]
Linter report from 2023-11-06 17:56, for mini language OCaml_BidirectionalRecords
```
File '_build/default/lib/tests.ml' doesn't have corresponding .mli interface
File '_build/default/lib/parser.ml' doesn't have corresponding .mli interface

File "lib/tests.ml", line 1, characters 0-13:
1 | open Angstrom
    ^^^^^^^^^^^^^
Alert zanuda-linter:  First item in file should be a documentation comment with copyright information. For example:
            (** Copyright 2021-2022, Winnie Pooh et al. *)
File "lib/tests.ml", line 2, characters 0-11:
2 | open Parser
    ^^^^^^^^^^^
Alert zanuda-linter: Second item in file should be a documentation comment with correct license information. For example:
           (** SPDX-License-Identifier: LGPL-3.0-or-later *)
File "lib/tests.ml", line 9, characters 17-25:
9 |   | Error msg -> failwith msg
                     ^^^^^^^^
Alert zanuda-linter: Using `failwith` (or `assert false`) usually is a clue that a corner case is not being handled properly. To report errors we recommend using error monad instead. In princliple, these construction are OK for temporary work-in-progress code, but in release they should be eliminated
File "lib/ast.ml", line 1, characters 0-34:
1 | type id = string [@@deriving show]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Alert zanuda-linter:  First item in file should be a documentation comment with copyright information. For example:
            (** Copyright 2021-2022, Winnie Pooh et al. *)
File "lib/ast.ml", line 1, characters 0-34:
1 | type id = string [@@deriving show]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Alert zanuda-linter: Second item in file should be a documentation comment with correct license information. For example:
           (** SPDX-License-Identifier: LGPL-3.0-or-later *)
```

File: OCaml_BidirectionalRecords/lib/ast.ml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  | NonRec
+[@@deriving show]
+
+type expr =
+  | EVar of ty_bind
+  | EConst of const
+  | EBinOp of binop * expr * expr (** EBinOp (op,l,r) is l op r *)
+  | EUnOp of unop * expr (** EUnOp (op,e) is op e *)
+  | EFun of ty_bind * expr (** Fun x -> e *)
+  | ECons of expr * expr (** ECons (h,t) is list h::t *)
```
Comment (Kakadu) [inline OCaml_BidirectionalRecords/lib/ast.ml:69]: А пустой где?
Кто-то ещё уже делал такую ошибку....

File: OCaml_BidirectionalRecords/lib/ast.ml [hunk 2]
```diff
@@ -0,0 +1,78 @@
+type id = string [@@deriving show]
+
+type ty =
+  | TInt
+  | TBool
+  | TString
+  | TList of ty
+  | TTuple of ty list
+  | Unspecified
```
Comment (Kakadu) [inline OCaml_BidirectionalRecords/lib/ast.ml:9]: Кринж. Такое вроде уже было у Ёлкина.

File: OCaml_BidirectionalRecords/lib/ast.ml (hunk 1 above)
Comment (Kakadu) [reply OCaml_BidirectionalRecords/lib/ast.ml:69]: А, вы его в const запихнули....

File: OCaml_BidirectionalRecords/lib/ast.ml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  | EBinOp of binop * expr * expr (** EBinOp (op,l,r) is l op r *)
+  | EUnOp of unop * expr (** EUnOp (op,e) is op e *)
+  | EFun of ty_bind * expr (** Fun x -> e *)
+  | ECons of expr * expr (** ECons (h,t) is list h::t *)
+  | ETuple of expr list (** (expr1, ..., exprn) *)
+  | EIfThenElse of expr * expr * expr (** IfThenElse (b,t,e) is if b then t else e *)
+  | ELet of rec_flag * ty_bind * expr * expr (** Let (x,e,e') is let x = e in e' *)
+  | EApp of expr * expr (**  App (f,e) is application f e *)
+  | EMatch of expr * (pattern * expr)
+  | EUnit
```
[the remark this answers is not in the document]
Comment (ksenmel) [reply OCaml_BidirectionalRecords/lib/ast.ml]: В данный момент это нужно для парсинга let, хочу позже потом исправить

**Comment by Kakadu** [comment]
Документация и тестовое покрытие  должны скоро появиться.

https://kakadu.github.io/fp2023/docs/OCaml_BidirectionalRecords

https://kakadu.github.io/fp2023/cov/OCaml_BidirectionalRecords

2023-11-07 17:34

**Comment by Kakadu** [comment]
Linter report from 2023-11-07 17:35, for mini language OCaml_BidirectionalRecords
```
File '_build/default/lib/tests.ml' doesn't have corresponding .mli interface
File '_build/default/lib/parser.ml' doesn't have corresponding .mli interface

File "lib/tests.ml", line 13, characters 17-25:
13 |   | Error msg -> failwith msg
                      ^^^^^^^^
Alert zanuda-linter: Using `failwith` (or `assert false`) usually is a clue that a corner case is not being handled properly. To report errors we recommend using error monad instead. In princliple, these construction are OK for temporary work-in-progress code, but in release they should be eliminated
```

**Comment by Kakadu** [comment]
Думаю, остальные замечания Вы сможете починить в следующем PR

**Comment by Kakadu** [comment]
Документация и тестовое покрытие  должны скоро появиться.

https://kakadu.github.io/fp2023/docs/OCaml_BidirectionalRecords

https://kakadu.github.io/fp2023/cov/OCaml_BidirectionalRecords

2023-11-07 19:46

**Comment by Kakadu** [comment]
Linter report from 2023-11-07 19:47, for mini language OCaml_BidirectionalRecords
```
File '_build/default/lib/tests.ml' doesn't have corresponding .mli interface
File '_build/default/lib/parser.ml' doesn't have corresponding .mli interface

File "lib/tests.ml", line 13, characters 17-25:
13 |   | Error msg -> failwith msg
                      ^^^^^^^^
Alert zanuda-linter: Using `failwith` (or `assert false`) usually is a clue that a corner case is not being handled properly. To report errors we recommend using error monad instead. In princliple, these construction are OK for temporary work-in-progress code, but in release they should be eliminated
```
